### PR TITLE
Do-Not-Merge: Use default keychain to authenticate image pulls

### LINKED ISSUE #74: pack build fails to pull image when docker is configured to use gcr credential helper
Using own build of pack cli from master branch
(git sha: 1974cedf4fc4836b1df2c30b0cbcbce5dffbd2d3)

I can successfully build using builder from public repo on dockerhub
e.g. `pack build jldec/java-hello:v1 --builder projectriff/builder --path .`

But substituting for a builder which requires a gcr credential helper fails, even though `docker pull` of the same image succeeds.

```sh
$ echo $PFS_BUILDER_IMAGE
gcr.io/cf-spring-pfs-release/projectriff-builder-a527160a809f5f8d339af74b304e6c46:0.0.1-snapshot-ci-31a4547a8f07

$ pack build jldec/java-hello:vp1 --builder $PFS_BUILDER_IMAGE --path . 
2018/11/25 17:33:50 Using user provided builder image 'gcr.io/cf-spring-pfs-release/projectriff-builder-a527160a809f5f8d339af74b304e6c46:0.0.1-snapshot-ci-31a4547a8f07'
2018/11/25 17:33:50 Pulling builder image 'gcr.io/cf-spring-pfs-release/projectriff-builder-a527160a809f5f8d339af74b304e6c46:0.0.1-snapshot-ci-31a4547a8f07' (use --no-pull flag to skip this step)
Error: Error response from daemon: unauthorized: You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication

$ docker pull $PFS_BUILDER_IMAGE
0.0.1-snapshot-ci-31a4547a8f07: Pulling from cf-spring-pfs-release/projectriff-builder-a527160a809f5f8d339af74b304e6c46
Digest: sha256:03082ad86323f70a699c784ce8dc5afd9bc9554ba500f126bd7d8daee09a8eb5
Status: Image is up to date for gcr.io/cf-spring-pfs-release/projectriff-builder-a527160a809f5f8d339af74b304e6c46:0.0.1-snapshot-ci-31a4547a8f07
```

NOTE: I'm using the "standalone" gcr credential helper described [here](https://cloud.google.com/container-registry/docs/advanced-authentication).

```sh
gcloud components install docker-credential-gcr
docker-credential-gcr configure-docker
```

cc @scothis

### PR DESCRIPTION
**Pull request by fbiville**
Fixes #74.

### CODE REVIEW

**Comment by fbiville** [comment]
This is quite ugly but I've tested this changed successfully against a Docker image in a private repository and against GCR.

File: docker/docker.go [hunk 1]
```diff
@@ -59,7 +62,18 @@ func (d *Client) RunContainer(ctx context.Context, id string, stdout io.Writer,
...
-	rc, err := d.ImagePull(context.Background(), ref, dockertypes.ImagePullOptions{})
+	reference, _ := name.ParseReference(ref, name.WeakValidation)
+	authenticator, _ := authn.DefaultKeychain.Resolve(reference.Context().Registry)
+	encodedHeader, _ := authenticator.Authorization()
+	encodedToken := strings.Replace(encodedHeader, "Basic ", "", 1)
+	tokenBytes, _ := base64.StdEncoding.DecodeString(encodedToken)
+	tokenAtoms := strings.SplitN(string(tokenBytes), ":", 2)
+	rc, err := d.ImagePull(context.Background(), ref, dockertypes.ImagePullOptions{
+		RegistryAuth: base64.StdEncoding.EncodeToString([]byte(
+			fmt.Sprintf(`{"username": "%s", "password": "%s"}`,
```
Comment (fbiville) [inline docker/docker.go]: based on https://github.com/moby/moby/issues/34503#issuecomment-358253070

File: docker/docker.go [hunk 2]
```diff
@@ -59,7 +62,18 @@ func (d *Client) RunContainer(ctx context.Context, id string, stdout io.Writer,
 }
 
 func (d *Client) PullImage(ref string) error {
-	rc, err := d.ImagePull(context.Background(), ref, dockertypes.ImagePullOptions{})
+	reference, _ := name.ParseReference(ref, name.WeakValidation)
+	authenticator, _ := authn.DefaultKeychain.Resolve(reference.Context().Registry)
+	encodedHeader, _ := authenticator.Authorization()
+	encodedToken := strings.Replace(encodedHeader, "Basic ", "", 1)
```
Comment (fbiville) [inline docker/docker.go]: based on https://github.com/moby/moby/issues/34503#issuecomment-322246711

**Comment by fbiville** [comment]
This should not be merged as there is, I hope, a more robust approach or at least similar library code that does what this change does.

**Comment by ekcasey** [comment]
Our docker pull code moved to `buildpack/lifecycle`. I opened a PR with similar changes https://github.com/buildpack/lifecycle/pull/48. 

The implementation is very similar to the one you contributed.

**Comment by fbiville** [comment]
Great news, @ekcasey.
I suppose I can close this PR?

**Pull request closed without merging**